File: sdk/go/harness/aforge.go
package harness

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	defaultAforgeMaxConcurrent = 8
	defaultAforgeTimeout       = 1800
	aforgeLandingWindow        = 5
)

var (
	aforgeSemaphore chan struct{}
	aforgeSemOnce   sync.Once
)

// AforgeProvider invokes aforge's one-shot machine-readable harness mode.
type AforgeProvider struct {
	BinPath string
	runCLI  func(ctx context.Context, cmd []string, env map[string]string, cwd string, timeout, idleSeconds int, stdin []byte) (*CLIResult, error)
}

// NewAforgeProvider creates an Aforge provider. If binPath is empty, it
// defaults to "aforge".
func NewAforgeProvider(binPath string) *AforgeProvider {
	if binPath == "" {
		binPath = strings.TrimSpace(os.Getenv("AFORGE_BIN"))
		if binPath == "" {
			binPath = "aforge"
		}
	}
	return &AforgeProvider{BinPath: binPath, runCLI: runCLIWithStdinIdle}
}

func getAforgeSemaphore() chan struct{} {
	aforgeSemOnce.Do(func() {
		maxConcurrent := defaultAforgeMaxConcurrent
		if raw := strings.TrimSpace(os.Getenv("AFORGE_MAX_CONCURRENT")); raw != "" {
			if parsed, err := strconv.Atoi(raw); err == nil && parsed > 0 {
				maxConcurrent = parsed
			}
		}
		aforgeSemaphore = make(chan struct{}, maxConcurrent)
	})
	return aforgeSemaphore
}

func aforgeTimeout(options Options) int {
	if options.Timeout > 0 {
		return options.Timeout
	}
	if raw := strings.TrimSpace(os.Getenv("AGENTFIELD_HARNESS_TIMEOUT_SECONDS")); raw != "" {
		if parsed, err := strconv.Atoi(raw); err == nil && parsed > 0 {
			return parsed
		}
	}
	return defaultAforgeTimeout
}

func aforgeInnerTimeout(outer int) int {
	if outer > aforgeLandingWindow {
		return outer - aforgeLandingWindow
	}
	return 1
}

func stripOpenRouterPrefix(model string) string {
	return strings.TrimPrefix(model, "openrouter/")
}

func supportedAforgeVariant(variant string) (string, bool) {
	normalized := strings.ToLower(strings.TrimSpace(variant))
	switch normalized {
	case "off", "low", "medium", "high":
		return normalized, true
	default:
		return "", false
	}
}

func aforgeTaskInput(prompt, systemPrompt string) string {
	if systemPrompt = strings.TrimSpace(systemPrompt); systemPrompt != "" {
		return systemPrompt + "\n\nTask:\n" + prompt
	}
	return prompt
}

func aforgeCommand() (string, error) {
	command := strings.ToLower(strings.TrimSpace(os.Getenv("AGENTFIELD_AFORGE_COMMAND")))
	if command == "" {
		return "exec", nil
	}
	if command != "do" && command != "exec" {
		return "", fmt.Errorf("AGENTFIELD_AFORGE_COMMAND must be do or exec, got %q", command)
	}
	return command, nil
}

func parseAforgeEnvelope(stdout string) map[string]any {
	// Both canonical `do` and `exec` print one JSON object. Parse that shape
	// before the wrapper-compatible line scan.
	var envelope map[string]any
	if err := json.Unmarshal([]byte(strings.TrimSpace(stdout)), &envelope); err == nil {
		_, hasDeliverable := envelope["deliverable"]
		_, hasText := envelope["text"]
		if hasDeliverable || hasText {
			return envelope
		}
	}

	lines := strings.Split(stdout, "\n")
	for i := len(lines) - 1; i >= 0; i-- {
		line := strings.TrimSpace(lines[i])
		if line == "" {
			continue
		}
		envelope = nil
		if err := json.Unmarshal([]byte(line), &envelope); err != nil {
			continue
		}
		_, hasDeliverable := envelope["deliverable"]
		_, hasText := envelope["text"]
		if hasDeliverable || hasText {
			return envelope
		}
	}
	return nil
}

func aforgeNumber(value any) (float64, bool) {
	switch number := value.(type) {
	case float64:
		return number, true
	case float32:
		return float64(number), true
	case int:
		return float64(number), true
	case int64:
		return float64(number), true
	case json.Number:
		parsed, err := number.Float64()
		return parsed, err == nil
	default:
		return 0, false
	}
}

func aforgeCrashMessage(returnCode int, blockedOn, deliverable, stderr string) string {
	cleanStderr := StripANSI(strings.TrimSpace(stderr))
	exitContext := fmt.Sprintf("aforge exit code %d", returnCode)
	message := exitContext
	if returnCode < 0 {
		message = fmt.Sprintf("Process killed by signal %d. %s", -returnCode, exitContext)
	}
	switch {
	case cleanStderr != "":
		message += ". stderr: " + truncate(cleanStderr, 1000)
	case blockedOn != "":
		message += ". blocked_on: " + truncate(blockedOn, 1000)
	case deliverable != "":
		message += ". partial: " + truncate(deliverable, 1000)
	}
	return message
}

func (p *AforgeProvider) Execute(ctx context.Context, prompt string, options Options) (*RawResult, error) {
	sem := getAforgeSemaphore()
	select {
	case sem <- struct{}{}:
		defer func() { <-sem }()
	case <-ctx.Done():
		return nil, ctx.Err()
	}

	root := options.ProjectDir
	if root == "" {
		root = options.Cwd
	}
	if root == "" {
		root = "."
	}
	outerTimeout := aforgeTimeout(options)
	command, commandErr := aforgeCommand()
	if commandErr != nil {
		return &RawResult{
			IsError: true, ErrorMessage: commandErr.Error(), FailureType: FailureCrash,
			Metrics: Metrics{},
		}, nil
	}
	var cmd []string
	input := []byte(prompt)
	if command == "exec" {
		cmd = []string{
			p.BinPath, "exec", "--json", "-w", root,
			"--timeout", strconv.Itoa(aforgeInnerTimeout(outerTimeout)),
			"--context-fill", "60", "--completion-reserve", "65536",
		}
		if systemPrompt := strings.TrimSpace(options.SystemPrompt); systemPrompt != "" {
			cmd = append(cmd, "--system", systemPrompt)
		}
	} else {
		cmd = []string{
			p.BinPath, "do", "--json", "--yes-spend", "-w", root,
			"--timeout", strconv.Itoa(aforgeInnerTimeout(outerTimeout)),
		}
		input = []byte(aforgeTaskInput(prompt, options.SystemPrompt))
	}

	model, variant := options.resolveModelAndVariant()
	env := make(map[string]string)
	if command == "exec" {
		env["AFORGE_MODELS"] = ""
	}
	if model != "" {
		modelSlug := stripOpenRouterPrefix(model)
		env["AFORGE_MODEL"] = modelSlug
		if command == "exec" {
			cmd = append(cmd, "--model", modelSlug, "--plan-model", modelSlug)
		}
	}
	if normalized, ok := supportedAforgeVariant(variant); ok {
		env["AFORGE_EXEC_REASONING"] = normalized
	}
	// Caller-supplied environment wins over values derived from model/variant.
	for key, value := range options.Env {
		env[key] = value
	}

	started := time.Now()
	cliResult, err := p.runCLI(ctx, cmd, env, "", outerTimeout, 0, input)
	apiMS := int(time.Since(started).Milliseconds())
	if err != nil {
		if isExecNotFound(err) {
			return &RawResult{
				IsError:      true,
				ErrorMessage: fmt.Sprintf("AForge binary not found at '%s'. Install it with `af aforge ensure`, or set AFORGE_BIN to its path.", p.BinPath),
				FailureType:  FailureCrash,
				Metrics:      Metrics{DurationAPIMS: apiMS},
			}, nil
		}
		if strings.Contains(strings.ToLower(err.Error()), "timed out") || strings.Contains(strings.ToLower(err.Error()), "deadline exceeded") {
			return &RawResult{
				IsError:      true,
				ErrorMessage: err.Error(),
				FailureType:  FailureTimeout,
				Metrics:      Metrics{DurationAPIMS: apiMS},
			}, nil
		}
		return nil, err
	}

	envelope := parseAforgeEnvelope(cliResult.Stdout)
	resultText := ""
	blockedOn := ""
	stop := ""
	usage := map[string]any{}
	if envelope != nil {
		textKey := "deliverable"
		if command == "exec" {
			textKey = "text"
		}
		if text, ok := envelope[textKey].(string); ok {
			resultText = strings.TrimSpace(text)
		}
		if value, ok := envelope["blocked_on"].(string); ok {
			blockedOn = strings.TrimSpace(value)
		}
		if value, ok := envelope["usage"].(map[string]any); ok {
			usage = value
		}
		if command == "exec" {
			if value, ok := envelope["stop"].(string); ok {
				stop = strings.TrimSpace(value)
			}
		}
	}

	isError := cliResult.ReturnCode != 0 || resultText == "" || blockedOn != ""
	if command == "exec" {
		isError = cliResult.ReturnCode < 0 || resultText == "" ||
			(cliResult.ReturnCode != 0 && cliResult.ReturnCode != 2 && cliResult.ReturnCode != 3)
	}
	metrics := Metrics{DurationAPIMS: apiMS}
	if value, ok := aforgeNumber(usage["calls"]); ok {
		metrics.NumTurns = int(value)
	}
	if command == "exec" {
		if value, ok := aforgeNumber(envelope["turns"]); ok {
			metrics.NumTurns = int(value)
		}
	}
	if value, ok := aforgeNumber(usage["prompt_tokens"]); ok {
		metrics.InputTokens = int(value)
	}
	if value, ok := aforgeNumber(usage["completion_tokens"]); ok {
		metrics.OutputTokens = int(value)
	}
	if value, ok := aforgeNumber(usage["cached_tokens"]); ok {
		metrics.CacheReadTokens = int(value)
	}
	spend, hasSpend := aforgeNumber(envelope["spend"])
	legacyCost, hasLegacyCost := aforgeNumber(usage["cost"])
	if hasSpend && spend > 0 {
		cost := spend
		metrics.CostUSD = &cost
	} else if hasLegacyCost && legacyCost > 0 {
		cost := legacyCost
		metrics.CostUSD = &cost
	}

	messages := []map[string]any(nil)
	if envelope != nil {
		messages = []map[string]any{envelope}
	}
	raw := &RawResult{
		Result:      resultText,
		Messages:    messages,
		Metrics:     metrics,
		IsError:     isError,
		FailureType: FailureNone,
		ReturnCode:  cliResult.ReturnCode,
	}
	if isError {
		if (command == "do" && cliResult.ReturnCode == 2) || (command == "exec" && cliResult.ReturnCode == 4) {
			raw.FailureType = FailureTimeout
		} else {
			raw.FailureType = FailureCrash
		}
		raw.ErrorMessage = aforgeCrashMessage(cliResult.ReturnCode, firstNonEmpty(blockedOn, stop), resultText, cliResult.Stderr)
	}
	return raw, nil
}

File: sdk/typescript/src/harness/providers/aforge.ts
import type { HarnessProvider } from './base.js';
import { resolveRoot } from './base.js';
import type { RawResult } from '../types.js';
import { createMetrics, createRawResult } from '../types.js';
import { runCli } from '../cli.js';
import { resolveModelAndVariant } from '../modelVariant.js';

const REASONING_VARIANTS = new Set(['off', 'low', 'medium', 'high']);
const DEFAULT_TIMEOUT_SECONDS = 1800;
const LANDING_WINDOW_SECONDS = 5;
const DEFAULT_MAX_CONCURRENT = 8;
const ANSI_PATTERN = /\x1B\[[0-?]*[ -/]*[@-~]/g;

// From aforge's DefaultModel; keep in step with aforge's built-in default.
export const AFORGE_DEFAULT_MODEL = '~deepseek/deepseek-v4-flash-latest';

class Semaphore {
  private active = 0;
  private readonly waiters: Array<() => void> = [];

  public constructor(private readonly limit: number) {}

  public async use<T>(operation: () => Promise<T>): Promise<T> {
    await this.acquire();
    try {
      return await operation();
    } finally {
      this.release();
    }
  }

  private acquire(): Promise<void> {
    if (this.active < this.limit) {
      this.active += 1;
      return Promise.resolve();
    }
    return new Promise((resolve) => {
      this.waiters.push(() => {
        this.active += 1;
        resolve();
      });
    });
  }

  private release(): void {
    this.active -= 1;
    this.waiters.shift()?.();
  }
}

function resolveMaxConcurrent(): number {
  const parsed = Number.parseInt(process.env.AFORGE_MAX_CONCURRENT ?? '', 10);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : DEFAULT_MAX_CONCURRENT;
}

const aforgeSemaphore = new Semaphore(resolveMaxConcurrent());

function stripOpenRouterPrefix(model: string): string {
  return model.startsWith('openrouter/') ? model.slice('openrouter/'.length) : model;
}

function parseEnvelope(stdout: string): Record<string, unknown> | undefined {
  // Both canonical `do` and `exec` print one JSON object. Parse that shape
  // before the wrapper-compatible line scan.
  try {
    const value: unknown = JSON.parse(stdout.trim());
    if (typeof value === 'object' && value !== null && !Array.isArray(value)
      && ('deliverable' in value || 'text' in value)) {
      return value as Record<string, unknown>;
    }
  } catch {
    // Fall through to the wrapper-compatible line scan.
  }

  const lines = stdout.split('\n').map((line) => line.trim()).filter(Boolean);
  for (let index = lines.length - 1; index >= 0; index -= 1) {
    try {
      const value: unknown = JSON.parse(lines[index]);
      if (typeof value === 'object' && value !== null && !Array.isArray(value)
        && ('deliverable' in value || 'text' in value)) {
        return value as Record<string, unknown>;
      }
    } catch {
      // Tolerate stray stdout from a wrapper and keep looking for the envelope.
    }
  }
  return undefined;
}

function numeric(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined;
}

function timeoutSeconds(): number {
  const parsed = Number.parseInt(process.env.AGENTFIELD_HARNESS_TIMEOUT_SECONDS ?? '', 10);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : DEFAULT_TIMEOUT_SECONDS;
}

function innerTimeout(outer: number): number {
  return outer > LANDING_WINDOW_SECONDS ? outer - LANDING_WINDOW_SECONDS : 1;
}

function taskInput(prompt: string, systemPrompt: unknown): string {
  return typeof systemPrompt === 'string' && systemPrompt.trim()
    ? `${systemPrompt.trim()}\n\nTask:\n${prompt}`
    : prompt;
}

function crashMessage(exitCode: number, blockedOn: string, deliverable: string | undefined, stderr: string): string {
  const cleanStderr = stderr.trim().replace(ANSI_PATTERN, '');
  const exitContext = `aforge exit code ${exitCode}`;
  let message = exitCode < 0 ? `Process killed by signal ${-exitCode}. ${exitContext}` : exitContext;
  if (cleanStderr) {
    message += `. stderr: ${cleanStderr.slice(0, 1000)}`;
  } else if (blockedOn) {
    message += `. blocked_on: ${blockedOn.slice(0, 1000)}`;
  } else if (deliverable) {
    message += `. partial: ${deliverable.slice(0, 1000)}`;
  }
  return message;
}

function stringOptions(value: unknown): Record<string, string> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return {};
  }
  const result: Record<string, string> = {};
  for (const [key, item] of Object.entries(value)) {
    if (typeof item === 'string') {
      result[key] = item;
    }
  }
  return result;
}

/**
 * Aforge CLI provider. `exec` is the default direct one-shot entry point; set
 * `AGENTFIELD_AFORGE_COMMAND=do` to opt into Aforge's routed workflow.
 */
export class AforgeProvider implements HarnessProvider {
  private readonly bin: string;

  public constructor(bin = 'aforge') {
    this.bin = bin === 'aforge' ? (process.env.AFORGE_BIN?.trim() || bin) : bin;
  }

  public async execute(prompt: string, options: Record<string, unknown>): Promise<RawResult> {
    return aforgeSemaphore.use(() => this.executeImpl(prompt, options));
  }

  private async executeImpl(prompt: string, options: Record<string, unknown>): Promise<RawResult> {
    const root = resolveRoot(options) ?? '.';
    const outerTimeout = timeoutSeconds();
    const command = (process.env.AGENTFIELD_AFORGE_COMMAND ?? 'exec').trim().toLowerCase();
    if (command !== 'do' && command !== 'exec') {
      return createRawResult({
        isError: true,
        errorMessage: `AGENTFIELD_AFORGE_COMMAND must be 'do' or 'exec', got ${JSON.stringify(command)}`,
        failureType: 'crash',
        metrics: createMetrics(),
      });
    }
    const systemPrompt = options.systemPrompt ?? options.system_prompt;
    const cmd = command === 'exec'
      ? [
          this.bin,
          'exec',
          '--json',
          '-w',
          root,
          '--timeout',
          String(innerTimeout(outerTimeout)),
        ]
      : [
          this.bin,
          'do',
          '--json',
          '--yes-spend',
          '-w',
          root,
          '--timeout',
          String(innerTimeout(outerTimeout)),
        ];
    // --turns exists only on exec, not do. Aforge's --budget is a token budget,
    // not a USD cap, so maxBudgetUsd has no honest mapping.
    if (command === 'exec' && typeof options.maxTurns === 'number'
      && Number.isFinite(options.maxTurns) && options.maxTurns > 0) {
      cmd.push('--turns', String(Math.trunc(options.maxTurns)));
    }
    if (command === 'exec') {
      cmd.push('--context-fill', '60', '--completion-reserve', '65536');
    }
    if (command === 'exec' && typeof systemPrompt === 'string' && systemPrompt.trim()) {
      cmd.push('--system', systemPrompt.trim());
    }

    const { model, variant } = resolveModelAndVariant(options);
    const env: Record<string, string> = command === 'exec' ? { AFORGE_MODELS: '' } : {};
    if (model) {
      const slug = stripOpenRouterPrefix(model);
      env.AFORGE_MODEL = slug;
      if (command === 'exec') {
        cmd.push('--model', slug, '--plan-model', slug);
      }
    }
    if (variant) {
      const normalized = variant.trim().toLowerCase();
      if (REASONING_VARIANTS.has(normalized)) {
        env.AFORGE_EXEC_REASONING = normalized;
      }
    }
    Object.assign(env, stringOptions(options.env));

    const effectiveModel = model || env.AFORGE_MODEL || AFORGE_DEFAULT_MODEL;

    const startApi = Date.now();
    try {
      const { stdout, stderr, exitCode } = await runCli(cmd, {
        env,
        cwd: undefined,
        timeout: outerTimeout * 1000,
        idleSeconds: 0,
        inputText: command === 'exec' ? prompt : taskInput(prompt, systemPrompt),
      });
      const envelope = parseEnvelope(stdout);
      const outputValue = command === 'exec' ? envelope?.text : envelope?.deliverable;
      const resultText = typeof outputValue === 'string' && outputValue.trim()
        ? outputValue.trim()
        : undefined;
      const blockedOn = typeof envelope?.blocked_on === 'string' ? envelope.blocked_on.trim() : '';
      const stop = typeof envelope?.stop === 'string' ? envelope.stop.trim() : '';
      const usage = typeof envelope?.usage === 'object' && envelope.usage !== null && !Array.isArray(envelope.usage)
        ? envelope.usage as Record<string, unknown>
        : {};

      const isError = command === 'exec'
        ? exitCode < 0 || resultText === undefined || ![0, 2, 3].includes(exitCode)
        : exitCode !== 0 || resultText === undefined || blockedOn !== '';
      const inputTokens = Math.trunc(numeric(usage.prompt_tokens) ?? 0);
      const outputTokens = Math.trunc(numeric(usage.completion_tokens) ?? 0);
      const cacheReadTokens = Math.trunc(numeric(usage.cached_tokens) ?? 0);
      const calls = Math.trunc(numeric(command === 'exec' ? envelope?.turns : usage.calls) ?? 0);
      const nativeSpend = numeric(envelope?.spend);
      const legacyCost = numeric(usage.cost);
      const providerCost = nativeSpend !== undefined && nativeSpend > 0 ? nativeSpend : legacyCost;

      return createRawResult({
        result: resultText,
        messages: envelope ? [envelope] : [],
        metrics: createMetrics({
          durationApiMs: Date.now() - startApi,
          numTurns: calls,
          totalCostUsd: providerCost !== undefined && providerCost > 0 ? providerCost : undefined,
          usage,
          sessionId: '',
          inputTokens,
          outputTokens,
          cacheReadTokens,
          cacheCreationTokens: 0,
          totalTokens: inputTokens + outputTokens,
          model: effectiveModel,
        }),
        isError,
        errorMessage: isError ? crashMessage(exitCode, blockedOn || stop, resultText, stderr) : undefined,
        failureType: isError
          ? ((command === 'do' && exitCode === 2) || (command === 'exec' && exitCode === 4)
              ? 'timeout'
              : 'crash')
          : 'none',
        returnCode: exitCode,
      });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      if (message.includes('ENOENT')) {
        return createRawResult({
          isError: true,
          errorMessage: `AForge binary not found at '${this.bin}'. Install it with \`af aforge ensure\`, or set AFORGE_BIN to its path.`,
          failureType: 'crash',
          metrics: createMetrics({ durationApiMs: Date.now() - startApi }),
        });
      }
      const timedOut = /timed out|deadline exceeded|no progress/i.test(message);
      return createRawResult({
        isError: true,
        errorMessage: message,
        failureType: timedOut ? 'timeout' : 'crash',
        metrics: createMetrics({ durationApiMs: Date.now() - startApi }),
      });
    }
  }
}
